File: src/binary_tree/is_bst.java
package binary_tree;

import java.util.ArrayList;

class Solution11 {
  private static class Node {
    public int value;
    public Node left;
    public Node right;

    public Node(int data) {
      this.value = data;
    }
  }

  /**
   * 是否搜索二叉树 中序遍历
   * 
   * @param head
   * @return
   */
  public static boolean isBST1(Node head) {
    if (head == null)
      return true;
    ArrayList<Node> arr = new ArrayList<>();
    in(head, arr);
    for (int i = 1; i < arr.size(); i++) {
      if (arr.get(i).value <= arr.get(i - 1).value)
        return false;
    }
    return true;
  }

  private static void in(Node head, ArrayList<Node> arr) {
    if (head == null)
      return;
    in(head.left, arr);
    arr.add(head);
    in(head.right, arr);
  }

  public static boolean isBST2(Node head) {
    if (head == null)
      return true;
    return process(head).isBST;
  }

  private static class Info {
    public boolean isBST;
    public int max, min;

    public Info(boolean i, int max, int min) {
      isBST = i;
      this.max = max;
      this.min = min;
    }
  }

  private static Info process(Node x) {
    if (x == null)
      return null;

    Info lInfo = process(x.left);
    Info rInfo = process(x.right);

    // // 处理Info
    int max = x.value, min = x.value;

    if (lInfo != null) {
      max = Math.max(max, lInfo.max);
      min = Math.min(min, lInfo.min);
    }
    if (rInfo != null) {
      max = Math.max(max, rInfo.max);
      min = Math.min(min, rInfo.min);
    }

    boolean isBST = true;
    if (lInfo != null && !lInfo.isBST)
      isBST = false;
    if (rInfo != null && !rInfo.isBST)
      isBST = false;

    if (lInfo != null && lInfo.max >= x.value)
      isBST = false;
    if (rInfo != null && rInfo.min <= x.value)
      isBST = false;
    return new Info(isBST, max, min);
  }

  // ================== TEST ==================
  public static Node generateRandomBST(int maxLevel, int maxValue) {
    return generate(1, maxLevel, maxValue);
  }

  // for test
  public static Node generate(int level, int maxLevel, int maxValue) {
    if (level > maxLevel || Math.random() < 0.5) {
      return null;
    }
    Node head = new Node((int) (Math.random() * maxValue));
    head.left = generate(level + 1, maxLevel, maxValue);
    head.right = generate(level + 1, maxLevel, maxValue);
    return head;
  }

  public static void main(String[] args) {
    int maxLevel = 4;
    int maxValue = 100;
    int testTimes = 1000000;
    for (int i = 0; i < testTimes; i++) {
      Node head = generateRandomBST(maxLevel, maxValue);
      if (isBST1(head) != isBST2(head)) {
        System.out.println("Oops!");
      }
    }
    System.out.println("finish!");
  }
}

File: src/binary_tree/is_bst.cpp
#include <iostream>

using namespace std;

struct Node {
  int value;
  Node *left, *right;
  Node(int value) : value(value){};
};

class Solution {
 private:
  void in(Node *head, vector<Node *> &arr) {
    if (head == nullptr) return;
    in(head->left, arr);
    arr.push_back(head);
    in(head->right, arr);
  }

  struct Info {
    bool isBST;
    int max, min;
    Info(bool bst, int max, int min) : isBST(bst), max(max), min(min){};
  };

  Info *process(Node *x) {
    if (x == nullptr) return nullptr;
    Info *lInfo = process(x->left);
    Info *rInfo = process(x->right);

    int ma = x->value, mi = x->value;
    if (lInfo != nullptr) {
      ma = max(ma, lInfo->max);
      mi = min(mi, lInfo->min);
    }
    if (rInfo != nullptr) {
      ma = max(ma, rInfo->max);
      mi = min(mi, rInfo->min);
    }

    bool isBST = true;
    if (lInfo != nullptr && !lInfo->isBST) isBST = false;
    if (rInfo != nullptr && !rInfo->isBST) isBST = false;

    if (lInfo != nullptr && lInfo->max >= x->value) isBST = false;
    if (rInfo != nullptr && rInfo->min <= x->value) isBST = false;
    return new Info(isBST, ma, mi);
  }

 public:
  bool isBST1(Node *head) {
    if (head == nullptr) return true;
    vector<Node *> arr;
    in(head, arr);
    for (int i = 1; i < arr.size(); i++) {
      if (arr[i]->value <= arr[i - 1]->value) return false;
    }
    return true;
  }

  bool isBST2(Node *head) {
    if (head == nullptr) return true;
    return process(head)->isBST;
  }

  // ============== TEST ==============
  Node *generateRandomBST(int maxLevel, int maxValue) {
    return generate(1, maxLevel, maxValue);
  };
  Node *generate(int level, int maxLevel, int maxValue) {
    if (level > maxLevel || rand() < 0.5) {
      return nullptr;
    }
    Node *head = new Node((int)(rand() % maxValue));
    head->left = generate(level + 1, maxLevel, maxValue);
    head->right = generate(level + 1, maxLevel, maxValue);
    return head;
  }
};

int main() {
  Solution solution;
  int maxLevel = 4;
  int maxValue = 100;
  int testTimes = 1000000;
  for (int i = 0; i < testTimes; i++) {
    Node *head = solution.generateRandomBST(maxLevel, maxValue);
    if (solution.isBST1(head) != solution.isBST2(head)) {
      cout << "Oops!" << endl;
    }
  }
  cout << "finish!" << endl;
  return 0;
}